diff --git a/sed/execute.c b/sed/execute.c
index 485bca7..1cf9a47 100644
--- a/sed/execute.c
+++ b/sed/execute.c
@@ -592,6 +592,10 @@ open_next_file (const char *name, struct input *input)
 
       input_fd = fileno (input->fp);
       fstat (input_fd, &input->st);
+#ifdef __MVS__
+      int txtflag = input->st.st_tag.ft_txtflag;
+      int ccsid = input->st.st_tag.ft_ccsid;
+#endif
       if (!S_ISREG (input->st.st_mode))
         panic (_("couldn't edit %s: not a regular file"), input->in_file_name);
 
@@ -621,6 +625,9 @@ open_next_file (const char *name, struct input *input)
       output_file.fp = ck_mkstemp (&input->out_file_name, tmpdir, "sed",
                                    write_mode);
       output_file.missing_newline = false;
+#ifdef __MVS__
+      __setfdccsid(fileno(output_file.fp), (txtflag << 16) | ccsid);
+#endif
       free (tmpdir);
 
       if (reset_fscreatecon)
@@ -658,6 +665,7 @@ closedown (struct input *input)
       target_name = input->in_file_name;
       input_fd = fileno (input->fp);
       output_fd = fileno (output_file.fp);
+      
 #ifdef HAVE_FCHOWN
       /* Try to set both UID and GID, but if that fails,
          try to set only the GID.  Ignore failure.  */
